src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts

```diff
@@ -231,4 +235,249 @@ describe('public-forms.verification.routes', () => {
+            hashRetries: 0,
+            hashedOtp: MOCK_HASHED_OTP,
+            signedData: MOCK_SIGNED_DATA,
+          }),
+        ],
+      })
+      const expectedResponse = {
+        message: 'Your OTP has expired, please request for a new one.',
+      }
+
+      // Act
+      const response = await request
+        .post(
+          `/forms/${mockVerifiableFormId}/fieldverifications/${mockOtpExpiredTransaction._id}/fields/${mockEmailFieldId}/otp/verify`,
+        )
+        .send({ otp: MOCK_OTP })
+
+      // Assert
+      expect(response.status).toBe(StatusCodes.UNPROCESSABLE_ENTITY)
+      expect(response.body).toEqual(expectedResponse)
+    })
+
+    it('should return 422 when the OTP retry limit is exceeded', async () => {
+      // Arrange
+      const requestForOtp = () =>
+        request
+          .post(
+            `/forms/${mockVerifiableFormId}/fieldverifications/${mockTransactionId}/fields/${mockEmailFieldId}/otp/verify`,
+          )
+          .send({ otp: MOCK_OTP })
+      const expectedResponse = {
+        message:
+          'You have entered too many invalid OTPs. Please request for a new OTP and try again.',
+      }
+
+      // Act
+      await Promise.allSettled(
+        _.range(NUM_OTP_RETRIES).map(() => requestForOtp()),
+      )
+      const response = await requestForOtp()
+
+      // Assert
+      expect(response.status).toBe(StatusCodes.UNPROCESSABLE_ENTITY)
+      expect(response.body).toEqual(expectedResponse)
+    })
+
+    it('should return 422 when the otp is wrong', async () => {
+      // Arrange
+      jest.spyOn(bcrypt, 'compare').mockResolvedValueOnce(false)
+      const expectedResponse = {
+        message: 'Wrong OTP.',
+      }
+
+      // Act
+      const response = await request
+        .post(
+          `/forms/${mockVerifiableFormId}/fieldverifications/${mockTransactionId}/fields/${mockEmailFieldId}/otp/verify`,
+        )
+        .send({ otp: '000000' })
+
+      // Assert
+      expect(response.status).toBe(StatusCodes.UNPROCESSABLE_ENTITY)
+      expect(response.body).toEqual(expectedResponse)
+    })
+
+    it('should return 500 when hashing error occurs', async () => {
+      // Arrange
+      jest.spyOn(bcrypt, 'compare').mockRejectedValueOnce(false)
+      const expectedResponse = {
+        message: 'Sorry, something went wrong. Please refresh and try again.',
+      }
+
+      // Act
+      const response = await request
+        .post(
+          `/forms/${mockVerifiableFormId}/fieldverifications/${mockTransactionId}/fields/${mockEmailFieldId}/otp/verify`,
+        )
+        .send({ otp: MOCK_OTP })
+
+      // Assert
+      expect(response.status).toBe(StatusCodes.INTERNAL_SERVER_ERROR)
+      expect(response.body).toEqual(expectedResponse)
+    })
+
+    it('should return 500 when database error occurs', async () => {
+      // Arrange
+      jest
+        .spyOn(VerificationModel, 'findById')
+        .mockReturnValue({ exec: jest.fn().mockRejectedValue('no') })
```

Comment:
Not sure, but is it okay that the `mockRejectedValue` is 'no' instead of an error?

Reply:
yeap! the reason this is ok (or not, because of `neverthrow` hehehe) is due to `neverthrow`'s  `ResultAsync.fromPromise`.

our calls to the database essentially returns `Promise<ModelSchema | null>`,  where the null is used to indicate an _expected_ error when something is off with the query and the requested information could not be found (eg: no model matching the specified options). 

however, promises and throwable functions actually have the possibility of an _exception_, which is an unexpected error that the called function isn't qualified to handle, and returns to teh caller. this is **not** encoded in the type signature and hence, is a mental overhead that engineers have to keep in the when calling functions, lest it crashes their system. In light of this, promises are actually typed as such: `Promise<your_return_type | exception>`

The `ResultAsync.fromPromise` function in `neverthrow` essentially allows us to sidestep this problem for promises (the corresponding function for throwables is `Result.fromThrowable`. This essentially allows us to catch a rejecting promise (or an error) and map it back using a function. Because the _type_ of exception is not encoded in ts' type system, this has to transform an unknown, back into a proper error type, which is done by `transformMongoError`

tl; dr: rejecting the promise causes it to go into error branch, and then a default error is returned from `transformMongoError` hence it's fine. thx for coming to my ted talk.

Reply:
Ohh okay thanks for sharing!

Reply:
@chowyiyin fyi the team decided a while ago that integration tests only need to cover Joi validation and success cases, since the controller unit tests should already test for the various error cases. @seaerchin has put in a lot of effort to mock lower layers even in integration tests so that error cases can be tested, which is great, but requires quite a bit of additional work.